dynamodb/schema.go

```diff
@@ -0,0 +1,367 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package dynamodb
+
+import (
+	"fmt"
+	"log"
+	"math/big"
+	"sort"
+
+	sp "cloud.google.com/go/spanner"
+	"github.com/aws/aws-sdk-go/aws"
+	"github.com/aws/aws-sdk-go/service/dynamodb"
+	"github.com/cloudspannerecosystem/harbourbridge/internal"
+	"github.com/cloudspannerecosystem/harbourbridge/schema"
+)
+
+const (
+	typeString          = "String"
+	typeBool            = "Bool"
+	typeNumber          = "Number"
+	typeNumberString    = "NumberString"
+	typeBinary          = "Binary"
+	typeList            = "List"
+	typeMap             = "Map"
+	typeStringSet       = "StringSet"
+	typeNumberSet       = "NumberSet"
+	typeNumberStringSet = "NumberStringSet"
+	typeBinarySet       = "BinarySet"
+
+	errThreshold      = float64(0.001)
+	conflictThreshold = float64(0.05)
+)
+
+type dynamoClient interface {
+	ListTables(input *dynamodb.ListTablesInput) (*dynamodb.ListTablesOutput, error)
+	DescribeTable(input *dynamodb.DescribeTableInput) (*dynamodb.DescribeTableOutput, error)
+	Scan(input *dynamodb.ScanInput) (*dynamodb.ScanOutput, error)
+}
+
+// ProcessSchema performs schema conversion for source tables in a DynamoDB
+// database. Since DynamoDB is a schemaless database, this process is imprecise.
+// We obtain schema information from two sources: from the table's metadata,
+// and from analyzing a sample of the table's rows.
+func ProcessSchema(conv *internal.Conv, client dynamoClient, tables []string, sampleSize int64) error {
+	if len(tables) == 0 {
+		var err error
+		tables, err = listTables(client)
+		if err != nil {
+			return err
+		}
+		if len(tables) == 0 {
+			return fmt.Errorf("no DynamoDB table exists under this account")
+		}
+	}
+	for _, t := range tables {
+		if err := processTable(conv, client, t, sampleSize); err != nil {
+			return err
+		}
+	}
+	schemaToDDL(conv)
+	conv.AddPrimaryKeys()
+	return nil
+}
+
+func listTables(client dynamoClient) ([]string, error) {
+	var tables []string
+	input := &dynamodb.ListTablesInput{}
+	for {
+		result, err := client.ListTables(input)
+		if err != nil {
+			return nil, err
+		}
+		for _, t := range result.TableNames {
+			tables = append(tables, *t)
+		}
+
+		if result.LastEvaluatedTableName == nil {
+			return tables, nil
+		}
+		input.ExclusiveStartTableName = result.LastEvaluatedTableName
+	}
+}
+
+func processTable(conv *internal.Conv, client dynamoClient, table string, sampleSize int64) error {
+	dySchema := dynamoDBSchema{TableName: table}
+	err := dySchema.analyzeMetadata(client)
+	if err != nil {
+		return err
+	}
+	stats, count, err := dySchema.scanSampleData(client, sampleSize)
+	if err != nil {
+		return err
+	}
+	dySchema.inferDataTypes(stats, count)
+	conv.SrcSchema[table] = dySchema.genericSchema()
+	return nil
+}
+
+type dynamoDBSchema struct {
```

Comment:
This datatype (along with colType and index) are very similar to schema.go.
Why not use schema.go directly?

Put another way, genericSchema() seems mostly a copy function - the only thing it seems to do is sort the list of column names. Could we just get rid of it?

Reply:
Fixed. This is a good suggestion. Looks much better now.